src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -199,44 +271,67 @@ public void PacketReset(SNIHandle handle, bool write, SNIPacket packet)
+            else if (serverNameParts.Length == 2 || serverNameParts.Length == 9)
             {
-                return CreateTcpHandle(serverNameParts[0], timerExpire, callbackObject, parallel);
-            }
+                int protocolLength = serverNameParts[0].Length + 1; // including ':'
+                string serverNameWithOutProtocol = fullServerName.Substring(protocolLength, fullServerName.Length - protocolLength);
 
-            switch (serverNameParts[0])
+                switch (serverNameParts[0])
+                {
+                    case TdsEnums.TCP:
+                        return CreateTcpHandle(serverNameWithOutProtocol, timerExpire, callbackObject, parallel, ref spnBuffer);
+
+                    case TdsEnums.NP:
+                        return CreateNpHandle(serverNameWithOutProtocol, timerExpire, callbackObject, parallel);
+
+                    default:
+                        if (parallel)
+                        {
+                            SNICommon.ReportSNIError(SNIProviders.INVALID_PROV, 0, SNICommon.MultiSubnetFailoverWithNonTcpProtocol, string.Empty);
+                        }
+                        else
+                        {
+                            SNICommon.ReportSNIError(SNIProviders.INVALID_PROV, 0, SNICommon.ProtocolNotSupportedError, string.Empty);
+                        }
+                        return null;
+                }
+            }
+            // all other cases are in wrong format
+            else
             {
-                case TdsEnums.TCP:
-                    return CreateTcpHandle(serverNameParts[1], timerExpire, callbackObject, parallel);
+                SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.INVALID_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
+                return null;
+            }
+        }
 
-                case TdsEnums.NP:
-                    return CreateNpHandle(serverNameParts[1], timerExpire, callbackObject, parallel);
+        private byte[] GetMsSqlServerSPN(string fullyQualifiedDomainName, int port = DefaultSqlServerPort)
+        {
+            string serverSpn = "MSSQLSvc/" + fullyQualifiedDomainName + ":" + port;
+            return Encoding.ASCII.GetBytes(serverSpn);
+        }
 
-                default:
-                    if (parallel)
-                    {
-                        SNICommon.ReportSNIError(SNIProviders.INVALID_PROV, 0, SNICommon.MultiSubnetFailoverWithNonTcpProtocol, string.Empty);
-                    }
-                    else
-                    {
-                        SNICommon.ReportSNIError(SNIProviders.INVALID_PROV, 0, SNICommon.ProtocolNotSupportedError, string.Empty);
-                    }
-                    return null;
-            }
+        private string GetFullyQualifiedDomainName(string hostNameOrAddress)
+        {
+            IPHostEntry hostEntry = Dns.GetHostEntry(hostNameOrAddress);
```

Comment:
This reverse lookup can potentially fail. Can you catch the exception if thrown and wrap it in a SqlException and provide an error message stating that Reverse lookup while generating the SPN has failed.